File: typescript-sdk/src/client/WorkflowClient.ts
import { HttpClient } from '../http/HttpClient';
import { 
  SysWorkflow, 
  WorkflowCreateRequest, 
  WorkflowUpdateRequest, 
  WorkflowExecutionRequest,
  PageResult,
  ExecutionStatus
} from '../types';
import { DataApiError } from '../errors';

/**
 * 工作流客户端
 */
export class WorkflowClient {
  private httpClient: HttpClient;
  
  constructor(httpClient: HttpClient) {
    this.httpClient = httpClient;
  }
  
  /**
   * 获取工作流列表
   */
  async getWorkflows(
    page: number = 0,
    size: number = 20,
    projectId?: string,
    userId?: string
  ): Promise<PageResult<SysWorkflow>> {
    try {
      const params: Record<string, any> = { page, size };
      if (projectId) params['projectId'] = projectId;
      if (userId) params['userId'] = userId;
      
      const response = await this.httpClient.get<PageResult<SysWorkflow>>('/workflows', params);
      return response.data;
    } catch (error) {
      throw new DataApiError('Failed to get workflows', 'GET_WORKFLOWS_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 根据ID获取工作流
   */
  async getWorkflow(id: string): Promise<SysWorkflow> {
    try {
      const response = await this.httpClient.get<SysWorkflow>(`/workflows/${id}`);
      return response.data;
    } catch (error) {
      throw new DataApiError(`Failed to get workflow ${id}`, 'GET_WORKFLOW_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 创建工作流
   */
  async createWorkflow(request: WorkflowCreateRequest): Promise<SysWorkflow> {
    try {
      const response = await this.httpClient.post<SysWorkflow>('/workflows', request);
      return response.data;
    } catch (error) {
      throw new DataApiError('Failed to create workflow', 'CREATE_WORKFLOW_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 更新工作流
   */
  async updateWorkflow(request: WorkflowUpdateRequest): Promise<SysWorkflow> {
    try {
      const response = await this.httpClient.put<SysWorkflow>(`/workflows/${request.id}`, request);
      return response.data;
    } catch (error) {
      throw new DataApiError(`Failed to update workflow ${request.id}`, 'UPDATE_WORKFLOW_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 删除工作流
   */
  async deleteWorkflow(id: string): Promise<void> {
    try {
      await this.httpClient.delete(`/workflows/${id}`);
    } catch (error) {
      throw new DataApiError(`Failed to delete workflow ${id}`, 'DELETE_WORKFLOW_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 检查工作流是否存在
   */
  async workflowExists(id: string): Promise<boolean> {
    try {
      const response = await this.httpClient.head(`/workflows/${id}`);
      return response.status === 200;
    } catch (error) {
      return false;
    }
  }
  
  /**
   * 执行工作流
   */
  async executeWorkflow(request: WorkflowExecutionRequest): Promise<{ executionId: string }> {
    try {
      const response = await this.httpClient.post<{ executionId: string }>('/workflows/execute', request);
      return response.data;
    } catch (error) {
      throw new DataApiError('Failed to execute workflow', 'EXECUTE_WORKFLOW_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 获取工作流执行状态
   */
  async getExecutionStatus(executionId: string): Promise<{ status: ExecutionStatus; result?: any; error?: string }> {
    try {
      const response = await this.httpClient.get<{ status: ExecutionStatus; result?: any; error?: string }>(
        `/workflows/executions/${executionId}/status`
      );
      return response.data;
    } catch (error) {
      throw new DataApiError(`Failed to get execution status ${executionId}`, 'GET_EXECUTION_STATUS_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 停止工作流执行
   */
  async stopExecution(executionId: string): Promise<void> {
    try {
      await this.httpClient.post(`/workflows/executions/${executionId}/stop`);
    } catch (error) {
      throw new DataApiError(`Failed to stop execution ${executionId}`, 'STOP_EXECUTION_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 获取工作流执行日志
   */
  async getExecutionLogs(executionId: string, page: number = 0, size: number = 100): Promise<PageResult<string>> {
    try {
      const response = await this.httpClient.get<PageResult<string>>(
        `/workflows/executions/${executionId}/logs`,
        { page, size }
      );
      return response.data;
    } catch (error) {
      throw new DataApiError(`Failed to get execution logs ${executionId}`, 'GET_EXECUTION_LOGS_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 获取工作流执行历史
   */
  async getExecutionHistory(
    workflowId: string,
    page: number = 0,
    size: number = 20
  ): Promise<PageResult<{ executionId: string; status: ExecutionStatus; startTime: string; endTime?: string }>> {
    try {
      const response = await this.httpClient.get<PageResult<{ executionId: string; status: ExecutionStatus; startTime: string; endTime?: string }>>(
        `/workflows/${workflowId}/executions`,
        { page, size }
      );
      return response.data;
    } catch (error) {
      throw new DataApiError(`Failed to get execution history for workflow ${workflowId}`, 'GET_EXECUTION_HISTORY_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 验证工作流定义
   */
  async validateWorkflow(definition: string): Promise<{ valid: boolean; errors?: string[] }> {
    try {
      const response = await this.httpClient.post<{ valid: boolean; errors?: string[] }>(
        '/workflows/validate',
        { definition }
      );
      return response.data;
    } catch (error) {
      throw new DataApiError('Failed to validate workflow', 'VALIDATE_WORKFLOW_ERROR', undefined, undefined, error as Error);
    }
  }
  
  /**
   * 复制工作流
   */
  async duplicateWorkflow(id: string, newName?: string): Promise<SysWorkflow> {
    try {
      const response = await this.httpClient.post<SysWorkflow>(
        `/workflows/${id}/duplicate`,
        newName ? { name: newName } : {}
      );
      return response.data;
    } catch (error) {
      throw new DataApiError(`Failed to duplicate workflow ${id}`, 'DUPLICATE_WORKFLOW_ERROR', undefined, undefined, error as Error);
    }
  }
}

File: java-sdk/src/main/java/com/owiseman/dataapi/sdk/client/WorkflowClient.java
package com.owiseman.dataapi.sdk.client;

import com.owiseman.dataapi.sdk.auth.AuthenticationProvider;
import com.owiseman.dataapi.sdk.config.ClientConfig;
import com.owiseman.dataapi.sdk.exception.DataApiException;
import com.owiseman.dataapi.sdk.model.*;
import com.owiseman.dataapi.sdk.http.HttpClient;
import com.fasterxml.jackson.core.type.TypeReference;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 工作流客户端
 */
public class WorkflowClient extends BaseClient {
    
    public WorkflowClient(ClientConfig config, AuthenticationProvider authProvider) {
        super(config, authProvider);
    }
    
    /**
     * 获取所有工作流
     */
    public List<SysWorkflow> getAllWorkflows(String projectId, String userId) throws DataApiException {
        Map<String, String> params = new HashMap<>();
        params.put("projectId", projectId);
        params.put("userId", userId);
        
        return httpClient.get("/api/workflows/findAll", params, new TypeReference<List<SysWorkflow>>() {});
    }
    
    /**
     * 分页获取工作流
     */
    public PageResult<SysWorkflow> getWorkflowsPage(String projectId, String userId, int page, int size) throws DataApiException {
        Map<String, String> params = new HashMap<>();
        params.put("projectId", projectId);
        params.put("userId", userId);
        params.put("page", String.valueOf(page));
        params.put("size", String.valueOf(size));
        
        return httpClient.get("/api/workflows/page", params, new TypeReference<PageResult<SysWorkflow>>() {});
    }
    
    /**
     * 根据ID获取工作流
     */
    public SysWorkflow getWorkflowById(String id) throws DataApiException {
        return httpClient.get("/api/workflows/" + id, null, SysWorkflow.class);
    }
    
    /**
     * 创建工作流
     */
    public SysWorkflow createWorkflow(WorkflowCreateRequest request) throws DataApiException {
        return httpClient.post("/api/workflows/create", request, SysWorkflow.class);
    }
    
    /**
     * 更新工作流
     */
    public SysWorkflow updateWorkflow(WorkflowUpdateRequest request) throws DataApiException {
        return httpClient.put("/api/workflows/update", request, SysWorkflow.class);
    }
    
    /**
     * 删除工作流
     */
    public void deleteWorkflow(String id) throws DataApiException {
        httpClient.delete("/api/workflows/" + id);
    }
    
    /**
     * 检查工作流是否存在
     */
    public boolean workflowExists(String id) throws DataApiException {
        try {
            httpClient.head("/api/workflows/" + id);
            return true;
        } catch (DataApiException e) {
            if (e.getStatusCode() == 404) {
                return false;
            }
            throw e;
        }
    }
    
    /**
     * 执行工作流
     */
    public WorkflowExecutionResult executeWorkflow(WorkflowExecutionRequest request) throws DataApiException {
        return httpClient.post("/api/workflow/execute", request, WorkflowExecutionResult.class);
    }
    
    /**
     * 获取工作流执行状态
     */
    public WorkflowExecutionStatus getExecutionStatus(String taskId) throws DataApiException {
        return httpClient.get("/api/workflow/status/" + taskId, null, WorkflowExecutionStatus.class);
    }
    
    /**
     * 停止工作流执行
     */
    public void stopExecution(String taskId) throws DataApiException {
        httpClient.post("/api/workflow/stop/" + taskId, null, Void.class);
    }
    
    /**
     * 获取工作流执行日志
     */
    public List<WorkflowExecutionLog> getExecutionLogs(String taskId) throws DataApiException {
        return httpClient.get("/api/workflow/logs/" + taskId, null, new TypeReference<List<WorkflowExecutionLog>>() {});
    }
}